pmm-app/src/pmm-qan-app-panel/QueryAnalytics/StateContext.tsx

```diff
@@ -0,0 +1,279 @@
+import React, { useEffect, useState } from 'react';
+import { ParseQueryParamDate } from '../../react-plugins-deps/components/helpers/time-parameters-parser';
+import { getDataSourceSrv } from '@grafana/runtime';
+import _ from 'lodash';
+
+const initialState = {} as any;
+
+export const StateContext = React.createContext(initialState);
+
+const DEFAULT_COLUMNS = ['load', 'num_queries', 'query_time'];
+const FILTERS_NAMES = [
+  'environment',
+  'cluster',
+  'replication_set',
+  'database',
+  'schema',
+  'node_name',
+  'service_name',
+  'client_host',
+  'username',
+  'service_type',
+  'node_type',
+  'city',
+  'az',
+];
+
+class ContextActions {
+  constructor() {}
+
+  static setFilters(query) {
+    return FILTERS_NAMES.reduce((result, filterName) => {
+      const filters = query.getAll(`var-${filterName}`);
+      if (!filters.length) {
+        return result;
+      }
+
+      if (filters[0] === 'All' || filters[0] === '') {
+        return result;
+      }
+
+      result[filterName] = filters;
+      return result;
+    }, {});
+  }
+
+  static refreshGrafanaVariables(state) {
+    const dataSource = getDataSourceSrv();
+    // @ts-ignore
+    const templateVariables = dataSource.templateSrv.variables;
+    FILTERS_NAMES.forEach(filter => {
+      const variables = _.find(templateVariables, { name: filter.replace('var-', '') });
+      if (!variables) {
+        return;
+      }
+      variables.current = {
+        text: state.labels[filter] || ['All'],
+        value: state.labels[filter] || ['All'],
+      };
+    });
+    templateVariables[0].variableSrv.variableUpdated(templateVariables[0]);
+  }
+
+  static generateURL(state) {
+    // read parameters and create new url
+    // @ts-ignore
+    const labels =
+      state.labels &&
+      Object.keys(state.labels)
```

Comment:
`Object.entries()` could be a better choice: it gives you both the key and the value